crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2352,16 +2352,107 @@ impl<'db> TypeInferenceBuilder<'db> {
         let ast::ExprCompare {
             range: _,
             left,
-            ops: _,
+            ops,
             comparators,
         } = compare;
 
-        self.infer_expression(left);
-        // TODO actually handle ops and return correct type
-        for right in comparators.as_ref() {
-            self.infer_expression(right);
+        // https://docs.python.org/3/reference/expressions.html#comparisons
+        // > Formally, if `a, b, c, …, y, z` are expressions and `op1, op2, …, opN` are comparison
+        // > operators, then `a op1 b op2 c ... y opN z` is equivalent to a `op1 b and b op2 c and
+        // ... > y opN z`, except that each expression is evaluated at most once.
+        let mut inferred_ty: Option<Type<'db>> = None;
+        std::iter::once(left.as_ref())
+            .chain(comparators.as_ref().iter())
+            // Evaluate expressions before iterating through pairs with `windows`
+            .map(|expr| self.infer_expression(expr))
+            // Allocation here is used to access the method `windows` not implemented on
+            // iterators (crates, experimental APIs or manual impl might provide it)
+            .collect::<Vec<Type<'db>>>()
+            .windows(2)
+            .zip(ops.iter())
+            .enumerate()
+            .for_each(|(i, (pair, op))| {
+                let (left_ty, right_ty) = (pair[0], pair[1]);
+                if inferred_ty.is_none() {
+                    let is_last = i == ops.len() - 1;
+                    let comparison_ty = self.infer_binary_type_comparison(left_ty, *op, right_ty);
+                    // Special case for `Unknown` returned when an operation is not supported
+                    inferred_ty = if let Type::Unknown = comparison_ty {
+                        Some(Type::Unknown)
+                    } else {
+                        match (comparison_ty.bool(self.db), is_last) {
+                            (Truthiness::AlwaysTrue, false) => None,
+                            (Truthiness::AlwaysTrue, true) => Some(Type::BooleanLiteral(true)),
+                            (Truthiness::AlwaysFalse, _) => Some(Type::BooleanLiteral(false)),
+                            (Truthiness::Ambiguous, _) => {
+                                Some(builtins_symbol_ty(self.db, "bool").to_instance(self.db))
+                            }
```

Comment:
Note that the `in`/`not in` operations always return bool though.

Reply:
Yes; also `is` and `not is`. But those are cases that will be reflected in the function that actually handles each binary comparison; it doesn't change the implications for this code, which needs to handle the results from any comparison.

Reply:
Hi guys, thank you so much for this very complete comment on the matter.

> It may make sense to try to actually extract and reuse the logic from our boolean and comparison, once we have here an iterator of types resulting from the comparisons that should be anded together? Or it may not, up to your judgment on that.

I actually started my PR with this approach. I realised that doing so (with my current rust mastery) led to additional allocations, and I know that a lot of effort went into avoiding allocations in the boolean PR, so I scrapped this approach. I think I will code my PR as a separate logic for now, and happily take suggestions on a direction to share the logic if one seems possible (either by accepting more allocations, or pointing me toward rust concepts that could make having both efficiency and code sharing possible).

> It will be a bit difficult to write tests showing that we're doing this correctly without supporting at least some form of comparison on Type::Instance (via looking up the return type of a comparison dunder method), since no built-in types return non-bools from comparison methods. It may make most sense to go ahead and implement the full dunder-method lookup logic (described at https://docs.python.org/3/reference/datamodel.html#object.__lt__ ), but if you can find a reasonable subset to implement with a TODO for the rest, that's fine too.

I think understanding thoroughly what to do here and writing tests will take me some time, but I'm happy to try. I will pass the PR as draft in the meantime.